File: JAVA/113 - Path Sum II.java
/*
 113. Path Sum II (M)

 Approach 1 : DFS
 Time: O(n)
 Space: O(n)

*/

/* Approach 1 */
class Solution {
    public List<List<Integer>> pathSum(TreeNode root, int sum) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        DFS(root, sum, new ArrayList<Integer>(), res);
        return res;
    }
    
    public void DFS(TreeNode root, int sum, List<Integer> cur, List<List<Integer>> res) {
        if (root == null) return;
        
        cur.add(root.val);
        
        if (root.left == null && root.right == null && root.val == sum) {
            res.add(new ArrayList<Integer>(cur));
        }
        
        DFS(root.left, sum - root.val, cur, res);
        DFS(root.right, sum - root.val, cur, res);
        cur.remove(cur.size() - 1);
    }
}

File: C++/113 - Path Sum II.cpp
/*
 113. Path Sum II (M)

 A1: Preorder (top-down) recursion
 Preorder the tree, use a vector to store the current path, if fullfill the
 given sum, save the path to result
 Time: O(n)
 Space: O(depth)

*/

class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int sum) {
        vector<vector<int>> res;
        vector<int> sub;
        
        pathSum(root, sum, sub, res);
        return res;
    }
    
    void pathSum(TreeNode* root, int sum, vector<int>& sub, vector<vector<int>>& res) {
        if (!root) return;
        
        // Add the current node in to path list
        sub.push_back(root->val);
        
        // Check if fullfill the given sum when it is a leaf node
        if (!root->left && !root->right && root->val == sum) {
            res.push_back(sub);
        }
        
        // If not, update the sum and move to the next level
        if (root->left) pathSum(root->left, sum - root->val, sub, res);
        if (root->right) pathSum(root->right, sum - root->val, sub, res);
        
        // Update the path for the next node
        sub.pop_back();
    }
};
